ssh-email-connect.man
.\" Man pages for ssh-email-connect

.TH man 1 "2013-11-01" "1.0" "ssh-email-connect man pages"

.SH NAME
.B "ssh-email-connect"

A wrapper around the ssh command that uses thunderbird inboxes to find the current IP address of a remote host. It uses information e-mailed from the remote host by boot-notify.service.

.SH SYNOPSIS
ssh-email-connect [-p PORT] [-u sshUSER] [-s SUBJECT] [-r remHOST] 

.SH "DESCRIPTION"
The ssh-email-connect script browses thunderbird inboxes searching for information from a remote host to find out the remote host's IP. The e-mail message from the remote hosts must have the host's IP as the first word. The boot-notify package provides a systemd service and scripts to send such e-mail messages. If ssh-email-connect finds such a message, the script ssh-connects the local user to the remote host.
.br

For safety reasons, the e-mail message from the remote host informs only about the host's current IP. The remote ssh port, user login and password should be known to the local user, or SSH key authentication should be set up and enabled for the two computers to connect.


.SH OPTIONS

.B -u sshUSER
.br
sshUSER is the user name/login that the local user will log in as, using the ssh network protocol. The default value of sshUSER is "$USER" (the current user on the local host).

.B -p PORT
.br
The ssh connection port. It defaults to 22 if not set otherwise.

.B -s SUBJECT
SUBJECT is the subject of the e-mail message ssh-email-connect looks for in thunderbird's inboxes.
.br
If not invoked explicitly, SUBJECT is assigned a default value of 'time msg from'. Not enough to differentiate e-mails from different hosts.

.B -r remHOST
.br
The remote host's name. If this option is used, " $remHOST" is appended to the $SUBJECT, and the script looks for e-mail subjects containing the value of "$SUBJECT $remHOST" (a space between $SUBJECT and $remHOST).


.SH AUTHOR's NOTES AND USAGE EXAMPLES
.br
(1) The remote host needs to send boot-time and shutdown notifications in a format recognizable by ssh-email-connect.
.br

(2) The boot-notify package/script provides a systemd service that sends the boot-time and shutdown notifications.
.br

(3) The ssh-email-connect script works only with thunderbird inboxes. Theoretically, it is possible to use other e-mail clients inboxes provided their structure is the same as thunderbird's.
.br

(4) The author has no plans to make the script work with other e-mail clients.
.br

(5) Use thunderbird's "Compact Folders" feature to avoid false positives. Note that a deleted e-mail message is not removed from an inbox until the "Compact Folders" command is invoked.
.br

.br
(6) The script could be made to work with profiles that held all the information necessary to connect to the remote hosts. In author's opinion that would be an overkill. Instead, the author suggests using aliases in .bashrc, e.g.

.br
	conn2neighbour='ssh-email-connect -p56991 -u xyz -r neighbour'
.br
	conn2mum='ssh-email-connect -u me -r mum'
.br
	conn2m='ssh-email-connect -u me -r m$'
.br

.br
The last example shows how to differentiate between the 'mum' and 'm' host names. Of course, in this example the remote host name must be the last word in the subject.
./"SEE ALSO
./"A list of related commands or functions.

.SH BUGS
.br
Not a bug, but be careful when you connect to multiple remote hosts. The default $SUBJECT will not be enough to differentiate e-mail messages from the different hosts.
.br
Also, see note (5) above.

.SH AUTHOR
.AU
Wiesław Magusiak <w.magusiak at gmail . com>

.SH COPYRIGHT
Do what you like with the ssh-email-connect, but do not blame me. No warranty is given, and no responsible is taken for the use.


ssh-email-connect.sh
#!/bin/bash
# By Wiesław Magusiak 
# SSH connection to a remote computer at an IP announced through e-mail
# Works only with thunderbird
#
(( $(which thunderbird 2>/dev/null 1>&2; echo $?) )) && { 
	echo "Missing dependency: Install thunderbird.";
	exit
}
VERSION=0.10
PORT=''
sshUSER=$USER
SUBJ="time msg from"

function id-inbox () {
	# Identify tunderbird's inbox containing a subject ($1)
	local SUBJ path0 path1 paths
	SUBJ="$1"
	path0="$HOME/.thunderbird"
	declare -a paths
	paths=$(grep Path= $path0/profiles.ini|cut -d= -f2 2>/dev/null)
	k=0 		# number of inboxes with $SUBJ found.

	for e in ${paths[@]}; do 
		path1="${path0}/${e}/Mail"
		for d in "${path1}"/* ; do
			if [[ -d $d ]]; then
				grep -m1 "$SUBJ" "${d}/Inbox" 1>/dev/null 2>&1; 
				if [[ $? == 0 ]]; then
					INBOX=$d
					((++k))
					[[ $k -gt 1 ]] && { echo "Error code 10"; return 10;}
				fi
			fi
		done
	done
	((k)) && echo $INBOX || { echo "Error code 11" && return 11;}
}

function myip () { 
	if [ -f `whereis curl | cut -d" " -f2` ] ; then 
		IP=$(curl -s ifconfig.me)
		IP=${IP#*: }; IP=${IP%%<*}
	else
		if [ -f `whereis wget | cut -d" " -f2` ] ; then 
			IP=$(wget -q -O - checkip.dyndns.org)
			IP=${IP#*: }; IP=${IP%%<*}
		else
			exit 1
		fi
	fi
	IFACE=$(echo $(ip route)|cut -d" " -f5)
	IP=$IP" "$(ip a show dev $IFACE | awk '$1 == "inet" { split($2, a, "/"); print a[1]; }')
	echo $IP" "$IFACE
}

# ------------- Start --------------------------------------------
if [[ $# > 0 ]] ; then
	USAGE="Usage:\n\t${0##*/} [-p PORT] [-u sshUSER] [-s SUBJECT] [-r remHOST]\
	\n\tDefault sshUSER='$sshUSER', PORT=22.\
	\n\tSee the man pages. (man ssh-email-connect)"
	while getopts  ":p:u:s:r:h" flag
	do
    	case "$flag" in
			p) PORT="-p$OPTARG";;
			u) sshUSER=$OPTARG;;
			s) SUBJ=$OPTARG;;
			r) remHOST=$OPTARG;;
			h) echo -e $USAGE && exit;;
    	esac
	done
fi
# ------------- Is the remote computer on? -----------------------
# E-mail inbox containing messages from a remote computer
[[ ${#remHOST} -gt 0 ]] && SUBJ=${SUBJ}" "${remHOST}
INBOX=$(id-inbox "$SUBJ")
if [[ ${INBOX%% *} == "Error" ]]; then
	echo -n "${INBOX}:  "
	[[ ${INBOX##* } == 11 ]] && echo "No e-mail received." || \
		echo "E-mail messages found in more than one inbox."
	exit ${INBOX##* }
fi
INBOX=$INBOX"/Inbox"

#remIP=$(grep -A7 "$SUBJ" $INBOX | tail -1)
remIP=$(echo $(grep -A12 "$SUBJ" $INBOX | tail -10|sed -n "/^$/,/^$/p"))

if [[ $remIP == 'Shutdown.' || $remIP == 'Close.' ]]; then
	echo The remote computer is switched off.
	exit 1
fi 

LOCAL=$(myip); 
[[ ${remIP%% *} != ${LOCAL%% *} ]] && remIP=${remIP%% *} || {
	remIP=${remIP% *}; remIP=${remIP#* };
	echo The remote computer is on the local network.
}


# ------------- The remote computer is switched on. -------------
echo -e "Connecting...\n\tssh $PORT $sshUSER@$remIP"
ssh $PORT $sshUSER@$remIP
